rust_src/src/filelock.rs

```diff
@@ -2,13 +2,277 @@
+
+        // The USER is everything before the last @.
+        let (user, rest) = data.split_at(data.rfind('@')?);
+        let user = user.to_string();
+
+        // The HOST is everything after the '@' to the last '.'.
+        let (at_host, rest) = rest.split_at(rest.rfind('.')?);
+        let host = at_host[1..].to_string();
+
+        // The PID is everything from the last '.' to the ':' or equivalent.
+        // NOTE: Unlike GNU Emacs, we do not tolerate negative PID values
+        // or over/underflow.
+        let (pid_str, boot_time_str) =
+            rest.split_at(rest.find(boot_time_sep).unwrap_or(rest.len()));
+        let pid = pid_str[1..].parse::<i32>().ok().filter(|pid| pid > &0)?;
+
+        // After the ':' or equivalent, if there is one, comes the boot time.
+        let boot_time = if boot_time_str.is_empty() {
+            None
+        } else {
+            // Invalid boot time results in a failed parse of the LockInfo
+            // rather than just boot_time = None.
+            // NOTE: Unlike GNU Emacs, this also applies to over/underflow.
+            // NOTE: Negative value is allowed.
+            Some(
+                boot_time_str
+                    .trim_start_matches(boot_time_sep)
+                    .parse()
+                    .ok()?,
+            )
+        };
+
+        Some(LockInfo {
+            user,
+            host,
+            pid,
+            boot_time,
+        })
+    }
+}
+
+/// The current locking state of a file.
+#[derive(PartialEq, Eq, Debug)]
+enum LockState {
+    /// The file is not locked.
+    NotLocked,
+    /// The file is locked by this Emacs process.
+    LockedByUs,
+    /// The file is locked by some other Emacs process identified
+    /// by [`LockInfo`].
+    LockedBy(LockInfo),
+}
+
+/// Opens a file in `O_NOFOLLOW` mode. Returns [`None`] if the open failed
+/// because the `path` refers to a symbolic link.
+#[cfg(unix)]
+fn open_nofollow(path: &Path) -> Result<Option<File>> {
+    use libc::{ELOOP, O_NOFOLLOW, O_RDONLY};
+    use std::ffi::CString;
+    use std::os::unix::ffi::OsStrExt;
+    use std::os::unix::io::FromRawFd;
+
+    let c_path =
+        CString::new(path.as_os_str().as_bytes()).map_err(|err| Error::new(InvalidInput, err))?;
+    let open_res = unsafe { emacs_open(c_path.as_ptr(), O_RDONLY | O_NOFOLLOW, 0) };
+
+    match open_res {
+        -1 => {
+            let os_error = Error::last_os_error();
+            if os_error.raw_os_error() == Some(ELOOP) {
+                Ok(None)
+            } else {
+                Err(os_error)
+            }
+        }
+
+        fd => {
+            let file = unsafe { File::from_raw_fd(fd) };
+            Ok(Some(file))
+        }
+    }
+}
+
+/// Reads the contents of a regular file at `path` up to `max_content_length` bytes.
+/// Returns [`None`] if the `path` is a symbolic link.
+fn read_nofollow(path: &Path, max_content_length: usize) -> Result<Option<String>> {
```

Comment:
I am on the fence. I think saving the result into a variable and returning `Ok(result)` makes it more clear that the only failures here are from the reading and parsing. 

```
let result = match open_nofollow(path)? {
    Some(file) => {
        let mut content = String::with_capacity(max_content_length);
            file.take(max_content_length as u64)
                .read_to_string(&mut content)?;
        Some(content)
    }
    None => None,
};
Ok(result)
```